lib/reek/spec/should_reek_of.rb

```diff
@@ -17,22 +23,42 @@ def initialize(smell_category,
 
       def matches?(actual)
         self.examiner = Examiner.new(actual, configuration: configuration)
-        self.all_smells = examiner.smells
-        all_smells.any? { |warning| warning.matches?(smell_category, smell_details) }
+
+        set_failure_messages_for_smell_type && return unless matching_smell_types?
+        set_failure_messages_for_smell_details && return unless matching_smell_details?
+
+        true
```

Comment:
OK, I see how the `case` does not show the strict dependency – but the two `&& return unless` lines didn’t either. :)

The clearest to me would probably be

```ruby
case
when matching_smell_types? and matching_smell_details?
  true
when matching_smell_types? and not matching_smell_details?
  set_failure_messages_for_smell_details
  false
when not matching_smell_types?
  set_failure_messages_for_smell_types
  false
end
```

…but I’m also OK with your `if`-`if`-`else`-`else` proposal.